src/NzbDrone.Core/Qualities/QualityDefinitionLimits.cs

```diff
@@ -0,0 +1,12 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+namespace NzbDrone.Core.Qualities;
+
+[SuppressMessage("Performance", "CA1822:Mark members as static", Justification =
+    "Required to be non-static for serialization purposes")]
+public record QualityDefinitionLimits
+{
+    public int MinLimit => 0;
+    public int PreferredLimit => 995;
```

Comment:
Could you help me understand the issue with preferred being an absolute number as opposed to an offset? So far in my coding and testing, it seems to work just fine. The distance between min & max is purely a UI concern, from what I can tell.

If the max value changes in the future, we simply adjust the preferred along with it as needed. Making it a distance means more math/work that clients have to repeat, which is one of the issues I'm trying to solve here.

Reply:
Thanks for the clarification.

> Because while preferred size has a finite maximum, the preferred size should never exceed the maximum allowed size for a quality. Currently the only thing that enforces this is the slider, the slider won't allow things to cross over or get within 5 points of each other. Likewise preferred shouldn't be allowed to be below the minimum size and the slider enforces that it should be 5 points above the minimum size.

I agree, and the API validation that I have yet to add to this PR should address this concern. That by itself doesn't require the API itself to be aware of any kind of relative offsets. Granted, I have not implemented the validation in terms of a specific distance from the min or max points, which I don't feel is necessary for the API side of things. Limiting the distance of preferred from its endpoints makes more sense in the UI for UX reasons; for example the user needs to be able to see & interact with the preferred circle control when it is really close to one of its endpoints, since if they're on top of each other one can't be clicked... this is why I was saying it's a UI concern.

> I can't think of a scenario where the preferred size should be less than the minimum or greater than the maximum, so enforcing the same behaviour as the slider on all inputs as well as in API validation.

I've removed all of my frontend changes, so whatever is being done there won't matter anymore for the purposes of this discussion. And in my text above I discuss validation on the API side, I explain that I don't see a reason for the API validation to necessarily need the same exact logic.

> Thinking about the UX, either the inputs would need to enforce those limits while changing or enforce the limits on the other values, such as pearling [zillow.github.io/react-slider](https://zillow.github.io/react-slider/)

I think with the focus now changed to the UI work to be done on the quality profiles page related to sizes, this is a separate concern. Obviously any UI validation/limits enforcement would benefit UX, but is no longer a hard requirement once API validation gets added. It'll be harder for not only the UI, but more importantly third party clients to submit invalid sizes.

I'll go ahead and resolve this conversation since this far into the discussion I still don't see a reason for the backend to concern itself with relative distance for preferred. As long as preferred <= max (for example), that's mathematically valid. Any stricter limits are up to the client (the UI in this scenario) to enforce. Again if I'm missing important functional considerations let me know, but otherwise I'll keep what I have.

Reply:
If the UI and API have different rules on what is valid then it'd be possible for an API client to set a value to something that isn't possible in the UI, while it's possible to correct that in the UI it's pretty janky and makes me wonder why enforce it in the API at all?

That said the more I think about it, the more I'd rather not have the difference at all, just enforce that Min <= Preferred and Max >= Preferred, but in that case I think the preferred maximum just disappears, since it doesn't need to be 5 less than the maximum, it just needs to not exceed the maximum limit or max size.

Reply:
> If the UI and API have different rules on what is valid then it'd be possible for an API client to set a value to something that isn't possible in the UI, while it's possible to correct that in the UI it's pretty janky and makes me wonder why enforce it in the API at all?

I agree with you, but that wasn't what I was trying to say. I apologize for the confusion; let me attempt to clarify. Let's say that the API and the UI both enforced the same rule. Let's just use min and max for example. So the rule would be `Min <= Max`. The UI-specific concern here is, how do we visually separate the two circle widgets so that the user can click each one? At a resolution of 1000, the spacing between say 999 and 1000 isn't very big. So we probably want some artificial gap that the slider uses to keep them separated enough for user interaction. This is the UI-only concern I was talking about. We might say that a spacing of 5 units is enough to get that spacing we want -- but that's not necessarily something the API enforces. The UI could, for example, do something like this:

```
slider.min = min(quality.min, quality.max - 5)
```

Users could still input a manual value that is closer than 5 units, but the slider will always render the circles a certain distance from each other. This means that if the user wants a high level of control or granularity, they probably need to stick with inputting manual numbers instead of using the slider itself.

This is just an example to demonstrate my point -- the UI can be *more restrictive* than the API, but in a way that doesn't really demand a difference in behavior from other clients, because the reasons for the stricter limit may not exist in those third party clients (e.g. UI elements).

> That said the more I think about it, the more I'd rather not have the difference at all, just enforce that Min <= Preferred and Max >= Preferred, but in that case I think the preferred maximum just disappears, since it doesn't need to be 5 less than the maximum, it just needs to not exceed the maximum limit or max size.

I agree here and this is the approach I was attempting to take. The more straightforward comparisons you mentioned are what I currently have in my validation rules. I also agree with your point about the preferred maximum. Removing this would also greatly simplify validation. I will go ahead and remove it in favor of the rule you described.

Reply:
I think even at a distance of 5 they become too close to really see which is which and part of the reason I'd like to come up with some different UX there, but I agree enforcing that gap at the API level becomes tricky and would be annoying to communicate clearly.